src/pyHepGrid/src/programs.py

```diff
@@ -671,3 +671,85 @@ class Sherpa(HEJ):
 
     _RUNFILE_END = ".dat"
     _WARMUP_FILES = ["Process", "Results.db"]
+
+class N3FIT(ProgramInterface):
+    """ Program interface for the n3fit fitting code """
+
+    def init_warmup(self, *args, **kwargs):
+        raise NotImplementedError("N3FIT does not implement any warmup mode")
+
+    def replica_folder(self, local_folder, runname = None):
+        if runname is None:
+            rep_folder = f"{local_folder}/nnfit"
+        else:
+            rep_folder = f"{local_folder}/{runname}/nnfit"
+        return rep_folder
+
+    def init_production(self, provided_warmup = None, continue_warmup = False, local = True):
+        """ Initialize production (single mode) """
+        if not local:
+            raise NotImplementedError("N3FIT is only implemented for local running")
+        _, expanded_card = util.expandCard()
+        for runcard, runfolder in expanded_card.items():
+            running_folder = self.get_local_dir_name(runcard, runfolder)
+            stdout_folder = f"{running_folder}/stdout"
+            # Create the folder in which the stuff will be run, and its stdout
+            if not util.checkIfThere(stdout_folder):
+                os.makedirs(stdout_folder)
+            # Now create also the replica folder to avoid future clashes
+            rep_folder = self.replica_folder(running_folder, runname = runfolder)
+            if not util.checkIfThere(rep_folder):
+                os.makedirs(rep_folder)
+            # And finally copy the runcard
+            from pyHepGrid.src.header import runcardDir
+            runcard_path = f"{runcardDir}/{runcard}"
+            from shutil import copy
+            copy(runcard_path, running_folder)
+        logger.info("run initialized")
+
+    def check_for_existing_output_local(self, runcard, runfolder, base_rep, n_reps):
+        """ Checks whether the given runcard already has the selected
+        replicas done.
+        For now miserably fail if that's indeed the case
+        """
+        from pyHepGrid.src.header import local_run_directory
+        running_folder = self.get_local_dir_name(runcard, runfolder)
+        replica_folder = self.replica_folder(running_folder, runname = runfolder)
+        # Use fitinfo file as a proxy for the replica
+        fitinfo = replica_folder + "/replica_{0}/" + runfolder + ".fitinfo"
+        for i in range(base_rep, base_rep + n_reps):
+            if util.checkIfThere(fitinfo.format(i)):
+                # Fail without piety
+                raise ValueError("There are replicas in the range considered")
+
+    def include_arguments(self, current_arguments):
```

Comment:
To check I understand - if you ever wanted to run NNPDF with Arc or DIRAC, would your `include_arguments` lead to command-line arguments for your executable like `--memsize X`?

@marianheil and I were discussing that supporting arbitrary arguments across the board might be quite useful,  especially if more programs are going to be added to `pyHepGrid`, so we don't have to add all programs' variables to all users' headers.  But it looks like they'd be ignored for SLURM, unless specifically added as a `{key}` to the `.sh` template, but for Arc/Dirac they'd be parsed and passed as additional `--X Y` command-line flags.

Ideally, for those programs we intend to be able to run on all the Backends, each runcard should lead to the same command-line arguments being passed regardless of which Backend it's being submitted to.

Any idea how we could both manage that, and arbitrary arguments?  (Probably related to #17 )

Reply:
> To check I understand - if you ever wanted to run NNPDF with Arc or DIRAC, would your `include_arguments` lead to command-line arguments for your executable like `--memsize X`?

Yes. Although it will fail because I'm  trusting in the existence of some arguments that are there only for slurm, but yes, that's the idea.

> @marianheil and I were discussing that supporting arbitrary arguments across the board might be quite useful, especially if more programs are going to be added to `pyHepGrid`, so we don't have to add all programs' variables to all users' headers. But it looks like they'd be ignored for SLURM, unless specifically added as a `{key}` to the `.sh` template, but for Arc/Dirac they'd be parsed and passed as additional `--X Y` command-line flags.

There are a few things that are going to be always backend dependent. In principle it might be possible to do it more general but disentangling all of that would be quite a task. I've tried in the past and got bored before achieving anything :__

In any case, this is particularly true for slurm. I remember we had some problems when giving options through command line arguments so we had to have them hard-coded on the scripts (which lead to having to have several different scripts). Don't remember whether this was on the IPPP batch.

> Any idea how we could both manage that, and arbitrary arguments? (Probably related to #17 )

I'd say #58 is not a bad compromise: each of the programs must decide/know how to treat the arguments.